node/src/network/p2p.rs

```diff
@@ -0,0 +1,188 @@
+use libp2p::{
+    core::muxing::StreamMuxerBox,
+    core::transport::Boxed,
+    futures::StreamExt,
+    identify, identity, kad,
+    mdns::Mdns,
+    noise,
+    swarm::{Swarm, SwarmEvent},
+    tcp, yamux, PeerId, Transport,
+};
+use std::error::Error;
+use std::time::Duration;
+use tokio::sync::mpsc;
+use tracing::{info, warn};
+
+pub struct P2PNetwork {
+    swarm: Swarm<ComposedBehaviour>,
+    command_receiver: mpsc::Receiver<NetworkCommand>,
+    event_sender: mpsc::Sender<NetworkEvent>,
+}
+
+#[derive(libp2p::swarm::NetworkBehaviour)]
+#[behaviour(out_event = "ComposedEvent")]
+struct ComposedBehaviour {
+    kad: kad::Behaviour<kad::store::MemoryStore>,
+    identify: identify::Behaviour,
+    mdns: Mdns,
+}
+
+enum ComposedEvent {
+    Kad(kad::Event),
+    Identify(identify::Event),
+    Mdns(mdns::Event),
+}
+
+pub enum NetworkCommand {
+    StartListening(String),
+    Dial(String),
+    AnnounceBlock(Vec<u8>),
+    AnnounceTransaction(Vec<u8>),
+}
+
+pub enum NetworkEvent {
+    NewPeer(PeerId),
+    PeerLeft(PeerId),
+    NewBlock(Vec<u8>),
+    NewTransaction(Vec<u8>),
+}
+
+impl P2PNetwork {
+    pub async fn new() -> Result<(Self, mpsc::Sender<NetworkCommand>, mpsc::Receiver<NetworkEvent>), Box<dyn Error>> {
```

Comment:
This composed result feels a bit odd IMO, been this the constructor of the P2PNetwork, but it returns even more stuff.